### PR TITLE
[#798] Save and list reports templates

### PR DESCRIPTION
**Pull request by danielkryska**
Preview: https://g3v8he.axshare.com/#g=1&p=scen1

- Add a button Zapisz jako szablon into Generuj raport modal on /map in zk options dropdown
  - Saved template will store the report title, the additional informations values (mini-map will be cropped), active layers, actual scene date as data źródła and active product, on loading set the latest scene
  - Show notification on correct save
- Add new modal Raporty that will contain a list of saved templates (only titles and descriptions)

Closes: #798.

### CODE REVIEW

**Review by kmarszalek — CHANGES_REQUESTED** [review]
code look ok, but after test on PR instance there is problem with loading report from list as described in slack.

**Review by kmarszalek — CHANGES_REQUESTED** [review]
once more, code looks good on backend side, front needs to be checked by Michal.
even with reports by savi on PR instance i wasnt able to make it work perfectly.
1. product will check but scene will or will not load on map -you never know
2. overlay will check and load on map - OK!
3. modal will pop up after selecting report- OK! but small map is wrong, it only has map, no overlay or scene.
PR needs polishing on front side.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ReportTemplateController.java
Comment (danielkryska) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ReportTemplateController.java:62]: Wouldn't be needed to add 403 error handling, as we will validate also beeing zk member in security?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java
Comment (danielkryska) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java:115]: Why you don't check is user is zk member?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java
Comment (danielkryska) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java:116]: Why you don't check is user is zk member?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java
Comment (danielkryska) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java:118]: Why you don't check is user is zk member?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ReportTemplateController.java
Comment (danielkryska) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ReportTemplateController.java:42]: Wouldn't be needed to add 403 error handling, as we will validate also beeing zk member in security?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateMapper.java
Comment (danielkryska) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateMapper.java:13]: What does this code? Is it only finding user by email? If yes, why isn't placed in AppUserService?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateService.java
Comment (danielkryska) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateService.java:40]: Whit method shouldn't be transactional?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateService.java
Comment (danielkryska) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateService.java:35]: Wouldn't we need some additional validation for products and overlays ids?

File: s4e-backend/src/test/java/pl/cyfronet/s4e/controller/ReportTemplateControllerTest.java
Comment (danielkryska) [inline s4e-backend/src/test/java/pl/cyfronet/s4e/controller/ReportTemplateControllerTest.java:250]: `shouldReturn403ForNonExistentReportTemplate`

Wouldn't it be `Existing`?

**Comment by danielkryska** [comment]
> once more, code looks good on backend side, front needs to be checked by Michal.
> even with reports by savi on PR instance i wasnt able to make it work perfectly.
> 
>     1. product will check but scene will or will not load on map -you never know
> 
>     2. overlay will check and load on map - OK!
> 
>     3. modal will pop up after selecting report- OK! but small map is wrong, it only has map, no overlay or scene.
>        PR needs polishing on front side.

I've made review and I'm concerned with a missing zk member validation fort the new endpoints.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateMapper.java
Comment (jswk) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateMapper.java:13]: It's the global MapStruct config actually.

The class, on the other hand, as it's name suggests, maps ReportTemplate transfer classes onto themselves. The one abstract method is in fact named incorrectly, it should be `createDtoToPreEntity` as the request is mapped by hand.
The user finding is a mapping operation, which fits nice in the mapper, simplifying the `ReportTemplateService`. This logic shouldn't be placed in `AppUserService` as in general services shouldn't use other services, except some corner cases, but refer directly to the repositories.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateService.java
Comment (jswk) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateService.java:40]: No

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateService.java
Comment (jswk) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/service/ReportTemplateService.java:35]: It's not crucial to keep these references correct if I understand the functionality correctly, so I don't validate them.

**Comment by jswk** [comment]
> I've made review and I'm concerned with a missing zk member validation fort the new endpoints.

Regarding the ZK member validation I didn't add it as this functionality may be exposed to any authenticated user. The same case is with SavedViews, which don't have this validation and are merely made inaccesible in the UI. @abacz what's your view on that?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ReportTemplateController.java
Comment (jswk) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ReportTemplateController.java:62]: See https://github.com/cyfronet-fid/sat4envi/pull/877#issuecomment-729529471.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java
Comment (jswk) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java:115]: See https://github.com/cyfronet-fid/sat4envi/pull/877#issuecomment-729529471.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java
Comment (jswk) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java:116]: See https://github.com/cyfronet-fid/sat4envi/pull/877#issuecomment-729529471.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java
Comment (jswk) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/config/SecurityConfig.java:118]: See https://github.com/cyfronet-fid/sat4envi/pull/877#issuecomment-729529471.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ReportTemplateController.java
Comment (jswk) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/controller/ReportTemplateController.java:42]: See https://github.com/cyfronet-fid/sat4envi/pull/877#issuecomment-729529471.

**Comment by jswk** [comment]
@danielkryska, as you're fixing your part don't amend my corrections from 624837b8.
I've recovered my commit and pushed it to the branch. Add your fixes in a new commit.

**Comment by jswk** [comment]
@danielkryska , I've again reverted your changes as you have overriden my changes. Add your changes in a separate commit and base it on the current HEAD of the branch, i.e. d1f6e1fd as I asked you in the previous comment https://github.com/cyfronet-fid/sat4envi/pull/877#issuecomment-729757375 and you seemed not to have any objections.

**Review by jswk — APPROVED** [review]
The behaviour now is much better, I didn't notice any glithes.

**Review by kmarszalek — APPROVED** [review]
now it works :)
wait for @michal-szostak  approval.

**Review by michal-szostak — APPROVED** [review]
I'm not 100% happy about the logic in front end about loading modal window after selecting report from list. But if @abacz is ok with this behavior I'm good.
The code itself is alright.

**Pull request merged**
